gnomad/utils/constraint.py

```diff
@@ -125,3 +136,251 @@ def collapse_strand(
+        "possible_variants": hl.agg.sum(all_high_coverage_ht.possible_variants),
+    }
+    for pop in pops:
+        agg_expr[f"observed_{pop}"] = hl.agg.array_sum(
+            all_high_coverage_ht[f"downsampling_counts_{pop}"]
+        )
+
+    # Generate a Table with all necessary annotations (x and y listed above) for the plateau models
+    high_coverage_ht = all_high_coverage_ht.group_by(*keys).aggregate(**agg_expr)
+    high_coverage_ht = annotate_mutation_type(high_coverage_ht)
+
+    # Build plateau models
+    plateau_models_agg_expr = build_plateau_models(
+        cpg_expr=high_coverage_ht.cpg,
+        mu_snp_expr=high_coverage_ht.mu_snp,
+        observed_variants_expr=high_coverage_ht.observed_variants,
+        possible_variants_expr=high_coverage_ht.possible_variants,
+        pop_observed_variants_exprs={
+            pop: high_coverage_ht[f"observed_{pop}"] for pop in pops
+        },
+        weighted=weighted,
+    )
+    plateau_models = high_coverage_ht.aggregate(hl.struct(**plateau_models_agg_expr))
+
+    # Filter to sites with coverage below `cov_cutoff` and larger than 0
+    all_low_coverage_ht = coverage_ht.filter(
+        (coverage_ht.exome_coverage < cov_cutoff) & (coverage_ht.exome_coverage > 0)
+    )
+
+    # Metric that represents the relative mutability of the exome calculated on high coverage sites
+    # This metric is used as scaling factor when building the coverage model
+    high_coverage_scale_factor = all_high_coverage_ht.aggregate(
+        hl.agg.sum(all_high_coverage_ht.observed_variants)
+        / hl.agg.sum(
+            all_high_coverage_ht.possible_variants * all_high_coverage_ht.mu_snp
+        )
+    )
+
+    # Generate a Table with all necessary annotations (x and y listed above) for the coverage model
+    low_coverage_ht = all_low_coverage_ht.group_by(
+        log_coverage=hl.log10(all_low_coverage_ht.exome_coverage)
+    ).aggregate(
+        low_coverage_obs_exp=hl.agg.sum(all_low_coverage_ht.observed_variants)
+        / (
+            high_coverage_scale_factor
+            * hl.agg.sum(
+                all_low_coverage_ht.possible_variants * all_low_coverage_ht.mu_snp
+            )
+        )
+    )
+
+    # Build the coverage model
+    # TODO: consider weighting here as well
+    coverage_model_expr = build_coverage_model(
+        low_coverage_obs_exp=low_coverage_ht.low_coverage_obs_exp,
+        log_coverage=low_coverage_ht.log_coverage,
+    )
+    coverage_model = tuple(low_coverage_ht.aggregate(coverage_model_expr).beta)
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
+    cpg_expr: hl.expr.BooleanExpression,
+    mu_snp_expr: hl.expr.Float64Expression,
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_exprs: List[hl.ArrayNumericExpression] = [],
+    weighted: bool = False,
+) -> Dict[str, Dict[bool, hl.expr.ArrayExpression]]:
```

Comment:
I found that the return type could be either `Dict[bool, hl.expr.ArrayExpression]` or `<ArrayExpression of type array<dict<bool, array<float64>>>>` when `pop_observed_variants_exprs` is specified. I explained it at the return type of the docstring. Should we specify this here as well?